File: 0785.判断二分图/0785-判断二分图.go
const UNCOLORED int = 0
const RED int = 1
const GREEN int = 2

func isBipartite(graph [][]int) bool {
    n := len(graph)
    color := make([]int, n)

    for i,_ := range graph {
        if color[i] == UNCOLORED {
            if !dfs( graph, color, i, RED ) {
                return false
            } 
        }
    }

    return true
}

func dfs(graph [][]int, color []int, index int, selfColor int) bool {
    color[index] = selfColor
    neighborColor := RED 
    if selfColor == RED {
        neighborColor = GREEN
    }
    
    for _, neighbor := range graph[index] {
        if color[neighbor] == UNCOLORED {
            if !dfs(graph, color, neighbor, neighborColor) {
                return false
            }
        } else if color[neighbor] != neighborColor {
            return false
        }
    }

    return true
}

File: 0785.判断二分图/0785-判断二分图.cpp
const int UNCOLORED = 0;
const int RED = 1;
const int GREEN = 2;

class Solution {
private:
    bool vaild;

public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        vaild = true;
        vector<int> colors(n, UNCOLORED);
        for( int i=0; i<n && vaild; i++ ) {
            if( colors[i] == UNCOLORED ) { // 只对未染色的节点操作;
                dfs(graph, colors, i, RED);
            }
        }

        return vaild;
    }

    void dfs(const vector<vector<int>>& graph, vector<int>& colors, int index, int c) {
        colors[index] = c;
        int neighborColor = ( c == RED ? GREEN:RED ) ;

        for( int neighbor : graph[index] ) {
            if( colors[neighbor] == UNCOLORED ) {
                dfs(graph, colors, neighbor, neighborColor);
            } else if( colors[neighbor] != neighborColor ) {
                vaild = false ;
                return;
            }
        } 

    }

};